RecoLocalCalo/HcalRecAlgos/src/HFSimpleTimeCheck.cc

```diff
@@ -0,0 +1,133 @@
+    eStates[1] = states[1];
+    if (!rejectAllFailures_)
+        for (unsigned i=0; i<2; ++i)
+            if (eStates[i] == HFAnodeStatus::FAILED_TIMING ||
+                eStates[i] == HFAnodeStatus::FAILED_OTHER)
+                eStates[i] = HFAnodeStatus::OK;
+    if (eStates[0] == HFAnodeStatus::OK)
+        return eStates[1];
+    else if (eStates[1] == HFAnodeStatus::OK)
+        return HFAnodeStatus::N_POSSIBLE_STATES + eStates[0] - 1;
+    else
+        return UINT_MAX;
+}
+
+HFRecHit HFSimpleTimeCheck::reconstruct(const HFPreRecHit& prehit,
+                                        const HcalCalibrations& /* calibs */,
+                                        const bool flaggedBadInDB[2])
+{
+    HFRecHit rh;
+
+    // Determine the status of each anode
+    unsigned states[2] = {HFAnodeStatus::NOT_READ_OUT, HFAnodeStatus::NOT_READ_OUT};
+    for (unsigned ianode=0; ianode<2; ++ianode)
+    {
+        const HFQIE10Info* anodeInfo = prehit.getHFQIE10Info(ianode);
+        if (anodeInfo)
+        {
+            if (flaggedBadInDB[ianode])
+                states[ianode] = HFAnodeStatus::FLAGGED_BAD;
+            else
+                states[ianode] = determineAnodeStatus(ianode, *anodeInfo);
+        }
+    }
+
+    // Reconstruct energy and time
+    const unsigned lookupInd = mapStatusIntoIndex(states);
+    if (lookupInd != UINT_MAX)
+    {
+        // In this scope, at least one of states[i] is HFAnodeStatus::OK
+        // or was mapped into that status by "mapStatusIntoIndex" method
+        //
+        const float* weights = &energyWeights_[lookupInd][0];
+        float energy = 0.f, t = 0.f, tfall = 0.f, weightedEnergySum = 0.f;
+        float tsum = 0.f, tfallsum = 0.f;
+        unsigned anodeCount = 0;
+
+        for (unsigned ianode=0; ianode<2; ++ianode)
+        {
+            const HFQIE10Info* anodeInfo = prehit.getHFQIE10Info(ianode);
+            if (anodeInfo)
+            {
+                const float weightedEnergy = weights[ianode]*anodeInfo->energy();
+                energy += weightedEnergy;
+                const float trising = anodeInfo->timeRising();
+                tsum += trising;
+                const float tfalling = anodeInfo->timeFalling();
+                tfallsum += tfalling;
+
+                if (weightedEnergy > 0.f)
+                {
+                    weightedEnergySum += weightedEnergy;
+                    t += trising*weightedEnergy;
+                    tfall += tfalling*weightedEnergy;
+                }
+
+                ++anodeCount;
+            }
+        }
+
+        if (weightedEnergySum > 0.f)
+        {
+            // Normally, determine TDC rise and fall time
+            // by using energy-weighted anode values
+            t /= weightedEnergySum;
+            tfall /= weightedEnergySum;
+        }
+        else
+        {
+            // But if energies are negative, use simple averages
+            t = tsum/anodeCount;
+            tfall = tfallsum/anodeCount;
```

Comment:
Already checked on line 81 and there is a comment
on lines 83 and 84 explaining what is going on.

On 06/14/2016 06:15 PM, Carl Vuosalo wrote:
>
> In RecoLocalCalo/HcalRecAlgos/src/HFSimpleTimeCheck.cc <https://github.com/cms-sw/cmssw/pull/14838#discussion_r67075125>:
>
> > +                ++anodeCount;
> > +            }
> > +        }
> > +
> > +        if (weightedEnergySum > 0.f)
> > +        {
> > +            // Normally, determine TDC rise and fall time
> > +            // by using energy-weighted anode values
> > +            t /= weightedEnergySum;
> > +            tfall /= weightedEnergySum;
> > +        }
> > +        else
> > +        {
> > +            // But if energies are negative, use simple averages
> > +            t = tsum/anodeCount;
> > +            tfall = tfallsum/anodeCount;
>
> If |anodeInfo| is null, |anodeCount| could be zero. Please check before dividing by it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/14838/files/94ddf556520df6348e6f7f6215e44341162a69a5#r67075125>, or mute the thread 
> <https://github.com/notifications/unsubscribe/AFS4-cMU7wCqaT-5DAleXCutddmxt0PPks5qLzYGgaJpZM4IyfiD>.
>